## 1. Import libraries

In [1]:
#import pandas
import pandas as pd

## 2. Import dataset

This dataset is retrieved from https://www.kaggle.com/datasets/elvinrustam/coursera-dataset?select=CourseraDataset-Unclean.csv. 

In [2]:
#Import dataset
coursera = pd.read_csv("CourseraDataset-Unclean.csv")

coursera.head()

Course Title  Rating           Level  \
0                                  Fashion as Design     4.8  Beginner level   
1                             Modern American Poetry     4.4  Beginner level   
2                          Pixel Art for Video Games     4.5  Beginner level   
3    Distribución digital de la música independiente     NaN  Beginner level   
4  The Blues: Understanding and Performing an Ame...     4.8  Beginner level   

                       Duration           Schedule         Review  \
0      20 hours (approximately)  Flexible schedule  2,813 reviews   
1  Approx. 34 hours to complete  Flexible schedule    100 reviews   
2       9 hours (approximately)  Flexible schedule    227 reviews   
3   Approx. 8 hours to complete  Flexible schedule            NaN   
4  Approx. 11 hours to complete  Flexible schedule    582 reviews   

                                 What you will learn  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Students will be able to describe the blues as...   

                                         Skill gain  \
0   ['Art History', 'Art', 'History', 'Creativity']   
1                                                []   
2                                                []   
3                                                []   
4  ['Music', 'Chord', 'Jazz', 'Jazz Improvisation']   

                                             Modules  \
0  ['Introduction', 'Heroes', 'Silhouettes', 'Cou...   
1  ['Orientation', 'Module 1', 'Module 2', 'Modul...   
2  ['Week 1: Introduction to Pixel Art', 'Week 2:...   
3   ['Semana 1', 'Semana 2', 'Semana 3', 'Semana 4']   
4  ['Blues Progressions – Theory and Practice ', ...   

                                          Instructor  \
0  ['Anna Burckhardt', 'Paola Antonelli', 'Michel...   
1                                    ['Cary Nelson']   
2             ['Andrew Dennis', 'Ricardo Guimaraes']   
3                      ['Eduardo de la Vara Brown.']   
4                              ['Dariusz Terefenko']   

                                       Offered By              Keyword  \
0                    ['The Museum of Modern Art']  Arts and Humanities   
1  ['University of Illinois at Urbana-Champaign']  Arts and Humanities   
2                   ['Michigan State University']  Arts and Humanities   
3                        ['SAE Institute México']  Arts and Humanities   
4                     ['University of Rochester']  Arts and Humanities   

                                          Course Url  
0      https://www.coursera.org/learn/fashion-design  
1  https://www.coursera.org/learn/modern-american...  
2  https://www.coursera.org/learn/pixel-art-video...  
3  https://www.coursera.org/learn/distribucion-di...  
4           https://www.coursera.org/learn/the-blues

## 3. Summarize Data

In [3]:
#Get the summarized information about the dataset
coursera.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9595 entries, 0 to 9594
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Course Title         9595 non-null   object 
 1   Rating               8156 non-null   float64
 2   Level                8330 non-null   object 
 3   Duration             9333 non-null   object 
 4   Schedule             8912 non-null   object 
 5   Review               8152 non-null   object 
 6   What you will learn  4984 non-null   object 
 7   Skill gain           9595 non-null   object 
 8   Modules              9595 non-null   object 
 9   Instructor           9595 non-null   object 
 10  Offered By           9595 non-null   object 
 11  Keyword              9595 non-null   object 
 12  Course Url           9595 non-null   object 
dtypes: float64(1), object(12)
memory usage: 974.6+ KB


There are 9595 rows x 13 cols.

*Duration* and *Review* columns should have Dtype **int** or **float**.

* **Course Title**:         No missing values. 
* **Rating**:               There are missing values.
* **Level**:                There are missing values. 
* **Duration**:             There are missing values. 
* **Schedule**:             There are missing values. 
* **Review**:               There are missing values. 
* **What you will learn**:  There are missing values. 
* **Skill gain**:           No missing values, "[]" can actually mean missing values.
* **Modules**:              No missing values, "[]" can actually mean missing values.
* **Instructor**:           No missing values, "[]" can actually mean missing values.
* **Offered By**:           No missing values, "[]" can actually mean missing values.
* **Keyword**:              No missing values.
* **Course Url**:           No missing values.  

## 4. Drop duplicates

In [4]:
#Drop duplicate rows from the dataset
coursera = coursera.drop_duplicates()

coursera.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8695 entries, 0 to 8694
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Course Title         8695 non-null   object 
 1   Rating               7346 non-null   float64
 2   Level                7589 non-null   object 
 3   Duration             8454 non-null   object 
 4   Schedule             8054 non-null   object 
 5   Review               7343 non-null   object 
 6   What you will learn  4585 non-null   object 
 7   Skill gain           8695 non-null   object 
 8   Modules              8695 non-null   object 
 9   Instructor           8695 non-null   object 
 10  Offered By           8695 non-null   object 
 11  Keyword              8695 non-null   object 
 12  Course Url           8695 non-null   object 
dtypes: float64(1), object(12)
memory usage: 951.0+ KB


There are 900 duplicates.

## 5. Inspect "Rating" column

A course on Coursera can have no rating if there is no user rating it. I decide to fill null values with 0.

In [5]:
#Fill null values with 0
coursera["Rating"] = coursera["Rating"].fillna(0)

#Get the summarized information about the Rating field
coursera["Rating"].describe()

count    8695.000000
mean        3.927683
std         1.697844
min         0.000000
25%         4.400000
50%         4.700000
75%         4.800000
max         5.000000
Name: Rating, dtype: float64

I want to check the data range of this field to make sure that all Rating values fall between 0.0 and 5.0. Checking the result above, we can see that the min is 0.0, and max is 5.0.

## 6. Inspect and modify "Duration" and "Review" columns

### 6.1. Null values of Duration column

In [6]:
#Check for null values of the Duration column
coursera[coursera["Duration"].isnull()]

Course Title  Rating  \
16        Create your UX portfolio with Adobe Portfolio     0.0   
235   Adobe Photoshop Para Principiantes: Efectos y ...     0.0   
398   Cómo crear historias para Instagram y Facebook...     4.2   
447   Edita tus fotos para las redes sociales con Easil     0.0   
448   Bearbeite deine Fotos für soziale Medien mit E...     0.0   
...                                                 ...     ...   
8637            Engage and Assess Students with Quizizz     4.7   
8655     Engaging and Assessing  Students with Plickers     4.7   
8659                     Use Buncee for Online Teaching     0.0   
8672     Crear contenidos para redes sociales con Prezi     0.0   
8683            Unleash Student Creativity with Buncee      0.0   

               Level Duration Schedule      Review  \
16    Beginner level      NaN      NaN         NaN   
235   Beginner level      NaN      NaN         NaN   
398   Beginner level      NaN      NaN  18 reviews   
447   Beginner level      NaN      NaN         NaN   
448   Beginner level      NaN      NaN         NaN   
...              ...      ...      ...         ...   
8637  Beginner level      NaN      NaN  24 reviews   
8655  Beginner level      NaN      NaN  35 reviews   
8659  Beginner level      NaN      NaN         NaN   
8672  Beginner level      NaN      NaN         NaN   
8683  Beginner level      NaN      NaN         NaN   

                                    What you will learn  \
16    In this project you will learn how to create a...   
235   Aplica filtros inteligentes a una imagen.Incor...   
398   En este proyecto aprenderás a utilizar Canva p...   
447   Crea un proyecto fotográfico individual en Eas...   
448   Designe dein individuelles Fotoprojekt für soz...   
...                                                 ...   
8637  Use Quizizz to curate and create quizzes for y...   
8655  Use Plickers to assess student understandingEn...   
8659  Use Buncee to create engaging learning activit...   
8672  Descubrirás las características de Prezi  Podr...   
8683  Create buncees to share science, literacy, and...   

                                             Skill gain  \
16    ['Share Content', 'Search Engine Optimization ...   
235   ['Adobe Photoshop', 'Photography', 'Photo Enha...   
398   ['Social Media Marketing', 'Social Media', 'Co...   
447   ['Social Media Marketing', 'Photography', 'Soc...   
448   ['Social Media Marketing', 'Photography', 'Gra...   
...                                                 ...   
8637  ['online tools', 'classroom tools', 'Teaching ...   
8655  ['online tools', 'student activities', 'studen...   
8659  ['interactive content', 'elementary teaching',...   
8672  ['Digital Media', 'Social Media Marketing', 'C...   
8683  ['online tools', 'communication tools', 'learn...   

                     Modules                       Instructor  \
16    ['Learn step-by-step']                ['Nicole Alvelo']   
235   ['Learn step-by-step']              ['Kevin Rodriguez']   
398   ['Learn step-by-step']  ['Monica I. Heredia Rodriguez']   
447   ['Learn step-by-step']              ['Michelle Flores']   
448   ['Learn step-by-step']              ['Michelle Flores']   
...                      ...                              ...   
8637  ['Learn step-by-step']               ['Rachel Lindsay']   
8655  ['Learn step-by-step']               ['Rachel Lindsay']   
8659  ['Learn step-by-step']               ['Rachel Lindsay']   
8672  ['Learn step-by-step']                ['Sara Nefzaoui']   
8683  ['Learn step-by-step']               ['Rachel Lindsay']   

                        Offered By              Keyword  \
16    ['Coursera Project Network']  Arts and Humanities   
235   ['Coursera Project Network']  Arts and Humanities   
398   ['Coursera Project Network']  Arts and Humanities   
447   ['Coursera Project Network']  Arts and Humanities   
448   ['Coursera Project Network']  Arts and Humanities   
...                            ...  

There are 241 null values in the *Duration* column. In many course links, the durations are shown in different languages, or units (such as minutes), or just "hour" without "approx.".

According to the uploader of this dataset, there are issues with the data scrapper that it's not being able to extract the values properly. Possible solutions would be:
* Drop the null values
* Modify data scrapper and merge with the missing data

In this project, I will drop the null values for the simplification purpose.

In [7]:
#Drop rows with null Duration values
coursera.drop(coursera[coursera["Duration"].isnull()].index, inplace = True)

coursera.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8454 entries, 0 to 8694
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Course Title         8454 non-null   object 
 1   Rating               8454 non-null   float64
 2   Level                7348 non-null   object 
 3   Duration             8454 non-null   object 
 4   Schedule             8048 non-null   object 
 5   Review               7256 non-null   object 
 6   What you will learn  4344 non-null   object 
 7   Skill gain           8454 non-null   object 
 8   Modules              8454 non-null   object 
 9   Instructor           8454 non-null   object 
 10  Offered By           8454 non-null   object 
 11  Keyword              8454 non-null   object 
 12  Course Url           8454 non-null   object 
dtypes: float64(1), object(12)
memory usage: 924.7+ KB


### 6.2. Convert Duration column
First is the *Duration* column, let's check which unit this column should have.

In [8]:
#Print out the number of times that each potential unit appears
print("Year:", coursera["Duration"].str.contains("year").sum(), "\n", 
      "Month:", coursera["Duration"].str.contains("month").sum(), "\n",
     "Week:", coursera["Duration"].str.contains("week").sum(), "\n",
     "Day:", coursera["Duration"].str.contains("day").sum(), "\n",
     "Hour:", coursera["Duration"].str.contains("hour").sum(), "\n",
     "Minute:", coursera["Duration"].str.contains("minute").sum(), "\n",
     "Second:", coursera["Duration"].str.contains("second").sum())

Year: 0 
 Month: 1621 
 Week: 1623 
 Day: 0 
 Hour: 8454 
 Minute: 15 
 Second: 0


From the result above, I will focus on converting these following units: Month, Week, Hour, and Minute. Because all values in the *Duration* column mention "hour", let's convert *Duration* into a float/int field representing the duration in hours.

In [9]:
#Make sure the units are standardized.
standard_unit = {"months":"month", "weeks":"week", "hours":"hour", "minutes":"minute"}

#Create a new data frame that includes the Course Title and Duration
duration = pd.DataFrame(coursera["Course Title"])
duration["duration_text"] = coursera["Duration"].replace(standard_unit, regex = True)

duration.head()

Course Title  \
0                                  Fashion as Design   
1                             Modern American Poetry   
2                          Pixel Art for Video Games   
3    Distribución digital de la música independiente   
4  The Blues: Understanding and Performing an Ame...   

                 duration_text  
0      20 hour (approximately)  
1  Approx. 34 hour to complete  
2       9 hour (approximately)  
3   Approx. 8 hour to complete  
4  Approx. 11 hour to complete

In [10]:
#Check for values that include "week" unit, but not "month".
duration[(duration["duration_text"].str.contains("week") == True) & 
         (duration["duration_text"].str.contains("month") == False)]

Course Title            duration_text
570  Improve Business Performance with Google Forms  1 week of study, 2 hour
574    Business Operations Support in Google Sheets  1 week of study, 2 hour

=> For this type of values, the approximate number of hours will be **the number of hours stated in the text**.

In [11]:
#Check for values that include "hour a week", "month", and exclude "minute" unit
duration[(duration["duration_text"].str.contains("hour a week") == True) & 
         (duration["duration_text"].str.contains("month") == True) & 
         (duration["duration_text"].str.contains("minute") == False)]

Course Title  \
30                        Graphic Design Specialization   
31                        UI / UX Design Specialization   
32                      Creative Writing Specialization   
33    Modern and Contemporary Art and Design Special...   
34         Game Design: Art and Concepts Specialization   
...                                                 ...   
8259  Creatividad, diseño e innovación: Técnicas y h...   
8261            Text Marketing Analytics Specialization   
8268  Liderazgo efectivo para el siglo XXI Specializ...   
8312  Preparing for Google Cloud Certification: Clou...   
8332  Teaching Impacts of Technology in K-12 Educati...   

                  duration_text  
30    2 month at 10 hour a week  
31    2 month at 10 hour a week  
32    1 month at 10 hour a week  
33    2 month at 10 hour a week  
34    1 month at 10 hour a week  
...                         ...  
8259  1 month at 10 hour a week  
8261  1 month at 10 hour a week  
8268  2 month at 10 hour a week  
8312  1 month at 10 hour a week  
8332  2 month at 10 hour a week  

[1621 rows x 2 columns]

=> There are 1621 rows that match these conditions, that means all rows with the "month" unit have this format: "*n* month at *m* hour a week". For this type of values, the approximate number of hours can be calculated as: **hour * 4 * month**

In [12]:
#Check for values with "minute" unit
duration[duration["duration_text"].str.contains("minute") == True]

Course Title         duration_text
543   Introduction to Business Analysis Using Spread...      1 hour 30 minute
2400                                      SQL Functions  1 hour and 30 minute
2636                Julia for Beginners in Data Science      1 hour 30 minute
4954                                     Networking 101  1 hour and 15 minute
4955                     Teaching with Google Classroom      1 hour 30 minute
5095  Using Ruby on Rails with Cloud SQL for Postgre...  1 hour and 20 minute
5172                      Containerization Using Docker      1 hour 15 minute
5188  Automate Doc Review Process with SharePoint & ...      1 hour 30 minute
5317      Process Forms with AI Model in Power Automate      1 hour 30 minute
5327             Getting Started with Cloud Data Fusion      1 hour 30 minute
5356  Distributed Machine Learning with Google Cloud ML      1 hour 30 minute
5455  Introduction to Business Analysis Using Spread...      1 hour 30 minute
5564  Machine Learning with Spark on Google Cloud Da...      1 hour 30 minute
5632  Logistic Regression for Classification using J...      1 hour 30 minute
6859     Process Personal Details using Methods in Java     2 hour, 15 minute

=> For this types of values, the approximate number of hours can be calculated as: **hour + minute/60**

Next, I can start working on extracting and calculating the number of hours.

In [13]:
#Split the string in the Duration field into a list of words
duration["duration_split"] = duration["duration_text"].str.split()

duration.head()

Course Title  \
0                                  Fashion as Design   
1                             Modern American Poetry   
2                          Pixel Art for Video Games   
3    Distribución digital de la música independiente   
4  The Blues: Understanding and Performing an Ame...   

                 duration_text                     duration_split  
0      20 hour (approximately)        [20, hour, (approximately)]  
1  Approx. 34 hour to complete  [Approx., 34, hour, to, complete]  
2       9 hour (approximately)         [9, hour, (approximately)]  
3   Approx. 8 hour to complete   [Approx., 8, hour, to, complete]  
4  Approx. 11 hour to complete  [Approx., 11, hour, to, complete]

In [14]:
#This function will try to convert string values into floats
def try_float (num):
    try:
        float(num)
    except:
        return 0
    else:
        return float(num)

In [15]:
#This function will calculate the number of hours stated in the Duration field
def get_hour (text):
    month = 0
    hour = 0
    minute = 0
    
    #Loop through the list to extract the numbers stand in front of the units
    for i in text:
        if (i == "month") & (text.index(i) != 0):
            month = try_float(text[text.index(i)-1])
        elif (i == "hour") & (text.index(i) != 0):
            hour = try_float(text[text.index(i)-1])
        elif (i == "minute") & (text.index(i) != 0):
            minute = try_float(text[text.index(i)-1])
    
    #Calculate the number of hours based on what I have identified previously
    approx_hours = hour
    
    if month != 0:
        approx_hours = approx_hours * month * 4

    approx_hours =  approx_hours + minute/60
    
    #Return with the calculated number of hours
    return approx_hours

In [16]:
#Apply the function to the duration_split column and create a new column to store the results
duration["approx_hours"] = duration["duration_split"].apply(get_hour)

duration.head()

Course Title  \
0                                  Fashion as Design   
1                             Modern American Poetry   
2                          Pixel Art for Video Games   
3    Distribución digital de la música independiente   
4  The Blues: Understanding and Performing an Ame...   

                 duration_text                     duration_split  \
0      20 hour (approximately)        [20, hour, (approximately)]   
1  Approx. 34 hour to complete  [Approx., 34, hour, to, complete]   
2       9 hour (approximately)         [9, hour, (approximately)]   
3   Approx. 8 hour to complete   [Approx., 8, hour, to, complete]   
4  Approx. 11 hour to complete  [Approx., 11, hour, to, complete]   

   approx_hours  
0          20.0  
1          34.0  
2           9.0  
3           8.0  
4          11.0

In [17]:
#Check for rows with approx_hours equals to 0
duration[duration["approx_hours"] == 0]

Course Title  \
752                            Business Model Canvas   
4888            Microsoft Azure Relational Databases   
6759  How To Create An Interactive Resume On Tableau   

                                          duration_text  \
752                                            one hour   
4888  This project should take approximately two hou...   
6759                                           1-2 hour   

                                         duration_split  approx_hours  
752                                         [one, hour]           0.0  
4888  [This, project, should, take, approximately, t...           0.0  
6759                                        [1-2, hour]           0.0

In [18]:
#Manually update the values based on the duration_text
duration.at[752,"approx_hours"] = 1
duration.at[4888, "approx_hours"] = 2
duration.at[6759, "approx_hours"] = 2

In [19]:
#Add the new column in the coursera dataframe
coursera["Duration (in hours, approximately)"] = duration["approx_hours"]

coursera.head()

Course Title  Rating           Level  \
0                                  Fashion as Design     4.8  Beginner level   
1                             Modern American Poetry     4.4  Beginner level   
2                          Pixel Art for Video Games     4.5  Beginner level   
3    Distribución digital de la música independiente     0.0  Beginner level   
4  The Blues: Understanding and Performing an Ame...     4.8  Beginner level   

                       Duration           Schedule         Review  \
0      20 hours (approximately)  Flexible schedule  2,813 reviews   
1  Approx. 34 hours to complete  Flexible schedule    100 reviews   
2       9 hours (approximately)  Flexible schedule    227 reviews   
3   Approx. 8 hours to complete  Flexible schedule            NaN   
4  Approx. 11 hours to complete  Flexible schedule    582 reviews   

                                 What you will learn  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Students will be able to describe the blues as...   

                                         Skill gain  \
0   ['Art History', 'Art', 'History', 'Creativity']   
1                                                []   
2                                                []   
3                                                []   
4  ['Music', 'Chord', 'Jazz', 'Jazz Improvisation']   

                                             Modules  \
0  ['Introduction', 'Heroes', 'Silhouettes', 'Cou...   
1  ['Orientation', 'Module 1', 'Module 2', 'Modul...   
2  ['Week 1: Introduction to Pixel Art', 'Week 2:...   
3   ['Semana 1', 'Semana 2', 'Semana 3', 'Semana 4']   
4  ['Blues Progressions – Theory and Practice ', ...   

                                          Instructor  \
0  ['Anna Burckhardt', 'Paola Antonelli', 'Michel...   
1                                    ['Cary Nelson']   
2             ['Andrew Dennis', 'Ricardo Guimaraes']   
3                      ['Eduardo de la Vara Brown.']   
4                              ['Dariusz Terefenko']   

                                       Offered By              Keyword  \
0                    ['The Museum of Modern Art']  Arts and Humanities   
1  ['University of Illinois at Urbana-Champaign']  Arts and Humanities   
2                   ['Michigan State University']  Arts and Humanities   
3                        ['SAE Institute México']  Arts and Humanities   
4                     ['University of Rochester']  Arts and Humanities   

                                          Course Url  \
0      https://www.coursera.org/learn/fashion-design   
1  https://www.coursera.org/learn/modern-american...   
2  https://www.coursera.org/learn/pixel-art-video...   
3  https://www.coursera.org/learn/distribucion-di...   
4           https://www.coursera.org/learn/the-blues   

   Duration (in hours, approximately)  
0                                20.0  
1                                34.0  
2                                 9.0  
3                                 8.0  
4                                11.0

### 6.3. Convert Review column
The *Review* column is quite similar to the *Rating* column: if there is no user reviews for the course, the value would be null.

First, let's make sure that all non-null values include "review" as the unit.

In [20]:
# Get the number of times that review unit appears
coursera["Review"].str.contains("reviews").sum()

7256

This equals the number of non-null values in the *Review* column. We will continue to extract numbers from the column and make sure that all non-null values of *Review* column has a matching value in the new column.

In [21]:
#Remove the comma from the text then extract number from the text
coursera["Number of reviews"] = coursera["Review"].str.replace(",","").str.extract('(\d+)')

#Check for rows that aren't able to extract numbers from non-null Review values.
coursera[coursera["Number of reviews"].isnull() & coursera["Review"].notnull()]

Empty DataFrame
Columns: [Course Title, Rating, Level, Duration, Schedule, Review, What you will learn, Skill gain, Modules, Instructor, Offered By, Keyword, Course Url, Duration (in hours, approximately), Number of reviews]
Index: []

That makes sure all Review values are transferred to the new Review column. Now, I will fill null values with "0" and change the dtype to float.

In [22]:
#Fill null values with "0" and change the dtype to float.
coursera["Number of reviews"] = coursera["Number of reviews"].fillna("0").astype(float)

### 6.4. Drop the old columns
We now can drop the old text fields of *Duration* and *Review*.

In [23]:
#Drop the old columns 
coursera.drop(columns=["Duration", "Review"], inplace = True)

coursera.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8454 entries, 0 to 8694
Data columns (total 13 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Course Title                        8454 non-null   object 
 1   Rating                              8454 non-null   float64
 2   Level                               7348 non-null   object 
 3   Schedule                            8048 non-null   object 
 4   What you will learn                 4344 non-null   object 
 5   Skill gain                          8454 non-null   object 
 6   Modules                             8454 non-null   object 
 7   Instructor                          8454 non-null   object 
 8   Offered By                          8454 non-null   object 
 9   Keyword                             8454 non-null   object 
 10  Course Url                          8454 non-null   object 
 11  Duration (in hours, approximately)  8454 no

## 7. Inspecting categorical fields: Level, Schedule, Keyword
These columns are categorical. Let's make sure that there are no typo in each categorical value.

In [24]:
#Get the unique values of Level
coursera["Level"].unique()

array(['Beginner level', 'Intermediate level', nan, 'Advanced level'],
      dtype=object)

In [25]:
#Get the unique values of Schedule
coursera["Schedule"].unique()

array(['Flexible schedule', nan], dtype=object)

In [26]:
#Get the unique values of Keyword
coursera["Keyword"].unique()

array(['Arts and Humanities', 'Business', 'Computer Science',
       'DataScience', 'Health', 'Information Technology',
       'Math and Logic', 'Personal Development',
       'Physical Science and Engineering', 'Social Sciences'],
      dtype=object)

There is no typo in these three fields. 

*Level* and *Schedule* have null values, I will fill those with "Not Specified"

In [27]:
#Fill null values with "Not Specified"
coursera[["Level","Schedule"]] = coursera[["Level","Schedule"]].fillna(value="Not Specified")

coursera.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8454 entries, 0 to 8694
Data columns (total 13 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Course Title                        8454 non-null   object 
 1   Rating                              8454 non-null   float64
 2   Level                               8454 non-null   object 
 3   Schedule                            8454 non-null   object 
 4   What you will learn                 4344 non-null   object 
 5   Skill gain                          8454 non-null   object 
 6   Modules                             8454 non-null   object 
 7   Instructor                          8454 non-null   object 
 8   Offered By                          8454 non-null   object 
 9   Keyword                             8454 non-null   object 
 10  Course Url                          8454 non-null   object 
 11  Duration (in hours, approximately)  8454 no

## 8. Clean other text fields
### 8.1. Fill null values of "What you will learn"

In [28]:
#Fill null values with "Not Specified"
coursera["What you will learn"] = coursera["What you will learn"].fillna(value = "Not Specified")

coursera.head()

Course Title  Rating           Level  \
0                                  Fashion as Design     4.8  Beginner level   
1                             Modern American Poetry     4.4  Beginner level   
2                          Pixel Art for Video Games     4.5  Beginner level   
3    Distribución digital de la música independiente     0.0  Beginner level   
4  The Blues: Understanding and Performing an Ame...     4.8  Beginner level   

            Schedule                                What you will learn  \
0  Flexible schedule                                      Not Specified   
1  Flexible schedule                                      Not Specified   
2  Flexible schedule                                      Not Specified   
3  Flexible schedule                                      Not Specified   
4  Flexible schedule  Students will be able to describe the blues as...   

                                         Skill gain  \
0   ['Art History', 'Art', 'History', 'Creativity']   
1                                                []   
2                                                []   
3                                                []   
4  ['Music', 'Chord', 'Jazz', 'Jazz Improvisation']   

                                             Modules  \
0  ['Introduction', 'Heroes', 'Silhouettes', 'Cou...   
1  ['Orientation', 'Module 1', 'Module 2', 'Modul...   
2  ['Week 1: Introduction to Pixel Art', 'Week 2:...   
3   ['Semana 1', 'Semana 2', 'Semana 3', 'Semana 4']   
4  ['Blues Progressions – Theory and Practice ', ...   

                                          Instructor  \
0  ['Anna Burckhardt', 'Paola Antonelli', 'Michel...   
1                                    ['Cary Nelson']   
2             ['Andrew Dennis', 'Ricardo Guimaraes']   
3                      ['Eduardo de la Vara Brown.']   
4                              ['Dariusz Terefenko']   

                                       Offered By              Keyword  \
0                    ['The Museum of Modern Art']  Arts and Humanities   
1  ['University of Illinois at Urbana-Champaign']  Arts and Humanities   
2                   ['Michigan State University']  Arts and Humanities   
3                        ['SAE Institute México']  Arts and Humanities   
4                     ['University of Rochester']  Arts and Humanities   

                                          Course Url  \
0      https://www.coursera.org/learn/fashion-design   
1  https://www.coursera.org/learn/modern-american...   
2  https://www.coursera.org/learn/pixel-art-video...   
3  https://www.coursera.org/learn/distribucion-di...   
4           https://www.coursera.org/learn/the-blues   

   Duration (in hours, approximately)  Number of reviews  
0                                20.0             2813.0  
1                                34.0              100.0  
2                                 9.0              227.0  
3                                 8.0                0.0  
4                                11.0              582.0

### 8.2. Convert lists into pure text strings
*Skill gain*, *Modules*, *Instructor*, *Offered By* are displayed as lists. I will convert them into strings and fill blank values with "Not Specified".

In [29]:
#Remove special characters "[", "]", and "'"
change_text = str.maketrans({"[":"", "]":"", "'":""})

coursera["Skill gain"] = coursera["Skill gain"].str.translate(change_text)
coursera["Modules"] = coursera["Modules"].str.translate(change_text)
coursera["Instructor"] = coursera["Instructor"].str.translate(change_text)
coursera["Offered By"] = coursera["Skill gain"].str.translate(change_text)

In [30]:
#Fill blank values with "Not Specified"
coursera[["Skill gain", "Modules", "Instructor", "Offered By"]] = coursera[["Skill gain", "Modules", "Instructor", "Offered By"]].replace("","Not Specified")

coursera.head()

Course Title  Rating           Level  \
0                                  Fashion as Design     4.8  Beginner level   
1                             Modern American Poetry     4.4  Beginner level   
2                          Pixel Art for Video Games     4.5  Beginner level   
3    Distribución digital de la música independiente     0.0  Beginner level   
4  The Blues: Understanding and Performing an Ame...     4.8  Beginner level   

            Schedule                                What you will learn  \
0  Flexible schedule                                      Not Specified   
1  Flexible schedule                                      Not Specified   
2  Flexible schedule                                      Not Specified   
3  Flexible schedule                                      Not Specified   
4  Flexible schedule  Students will be able to describe the blues as...   

                               Skill gain  \
0   Art History, Art, History, Creativity   
1                           Not Specified   
2                           Not Specified   
3                           Not Specified   
4  Music, Chord, Jazz, Jazz Improvisation   

                                             Modules  \
0  Introduction, Heroes, Silhouettes, Coutures, L...   
1  Orientation, Module 1, Module 2, Module 3, Mod...   
2  Week 1: Introduction to Pixel Art, Week 2: Pix...   
3             Semana 1, Semana 2, Semana 3, Semana 4   
4  Blues Progressions – Theory and Practice , Blu...   

                                          Instructor  \
0  Anna Burckhardt, Paola Antonelli, Michelle Mil...   
1                                        Cary Nelson   
2                   Andrew Dennis, Ricardo Guimaraes   
3                          Eduardo de la Vara Brown.   
4                                  Dariusz Terefenko   

                               Offered By              Keyword  \
0   Art History, Art, History, Creativity  Arts and Humanities   
1                           Not Specified  Arts and Humanities   
2                           Not Specified  Arts and Humanities   
3                           Not Specified  Arts and Humanities   
4  Music, Chord, Jazz, Jazz Improvisation  Arts and Humanities   

                                          Course Url  \
0      https://www.coursera.org/learn/fashion-design   
1  https://www.coursera.org/learn/modern-american...   
2  https://www.coursera.org/learn/pixel-art-video...   
3  https://www.coursera.org/learn/distribucion-di...   
4           https://www.coursera.org/learn/the-blues   

   Duration (in hours, approximately)  Number of reviews  
0                                20.0             2813.0  
1                                34.0              100.0  
2                                 9.0              227.0  
3                                 8.0                0.0  
4                                11.0              582.0

### 8.3. Confirm valid url
Last but not least, I want to make sure that the *Course Url* values are valid, meaning all values must at least contain “https://www.coursera.org/”

In [31]:
#Get the number of times that "https://www.coursera.org/" appears
coursera["Course Url"].str.contains("https://www.coursera.org/").sum()

8454

This means *Course Url* values are valid. The Data Cleaning process now is done.

## 9. Export the clean dataset

In [32]:
#Export the cleaned data into a csv file
coursera.to_csv("CourseraDataset-Clean.csv", index = False)